Editorial: Kite (#6331)
cpinitiative/usaco-guide: solutions/silver/cfgym-105822B.mdx

---
id: cfgym-105822B
source: CF
title: Kites
author: John Zhou
---

[Official Analysis](https://codeforces.com/gym/105822/attachments/download/31492/MITIT_2025_Spring_Editorials_1.pdf#page=2)

## Explanation

Sort the input array.
We must obtain $2$ pairs of the same number.

> To form a pair, it is always suboptimal to choose any two numbers that are not adjacent to each other after sorting.

Reviewer: why? little justification needed here about your observation.

Author: Done!

Thus, we compute $d[i]=a[i+1]-a[i]$, for $0<i<n-1$, and the problem reduces to choosing two $i$'s $i_1$ and $i_2$ such that $d[i_1]+d[i_2]$ is minimized.

> However, **the two pairs cannot share any number**, so the two pairs cannot be adjacent to each other, i.e. $i_1$ and $i_2$ cannot be adjacent.

Reviewer: ```suggestion
However, **the two pairs cannot share any number**, so the two pairs cannot be adjacent to each other, i.e., $i_1$ and $i_2$ cannot be adjacent.
```

Author: ```suggestion
However, **the two pairs cannot share any number**, so $i_1$ and $i_2$ cannot be adjacent to each other.
```

Extra cluase to get rid of

> However, **the two pairs cannot share any number**, so the two pairs cannot be adjacent to each other, i.e. $i_1$ and $i_2$ cannot be adjacent.

Reviewer: can you justify it with little example like:
pair1 used sticks at i1, i1 + 1
pair2 used at i2, i2 + 1
it's possible for i2 being equal to i1 + 1 then we are only using three sticks.

something like that

Author: Done!

Without loss of generality, $i_1 < i_2$.
We loop over $2\leq i_2<n-1$ (since $i_2=1$ is impossible), so now $d[i_2]$ is fixed;
$i_1$ must be in $1\leq i_1<i_2-1$, so we maintain a prefix minimum of $d[i_1]$ cumulatively to obtain the best $d[i_1]$ for each $i_2$.

## Implementation

> **Time Complexity:** $\mathcal{O}(N)$

Reviewer: ```suggestion
**Time Complexity:** $\mathcal{O}(N \log N)$
```

<LanguageSection>
<CPPSection>

```cpp
#include <algorithm>
#include <climits>
#include <iostream>
#include <vector>
using namespace std;

void solve() {
	int n;
	cin >> n;

> 	vector<int> a(n);

Reviewer: rename it to sticks

---
id: cfgym-105822B
source: CF
title: Kites
author: John Zhou
---

[Official Analysis](https://codeforces.com/gym/105822/attachments/download/31492/MITIT_2025_Spring_Editorials_1.pdf#page=2)

## Explanation

We must obtain $2$ pairs of the same number;
to make two indices $i_1$ and $i_2$ the same, $|a[i_1]-a[i_2]|$ operations are needed,

> so it is optimal to choose an adjacent pair of numbers in indices after sorting to make them the same.

Reviewer 2: ```suggestion
We must obtain $2$ pairs of equal numbers.
To make two indices $i_1$ and $i_2$ contain the same value, $|a[i_1]-a[i_2]|$ operations are needed.
Therefore, is is optimal to sort the array and make adjacent pair of numbers equal.
```
imo flows a lot better

Author: Agree but there was typo in suggestion.

```suggestion
We must obtain $2$ pairs of equal numbers.
To make two indices $i_1$ and $i_2$ contain the same value, $|a[i_1]-a[i_2]|$ operations are needed.
Therefore, it is optimal to sort the array and make adjacent pair of numbers equal.
```

Thus, we sort the input array, and compute $d[i]=a[i+1]-a[i]$ for $0<i<n-1$.

> The problem now reduces to choosing two $i$'s $i_1$ and $i_2$ such that $d[i_1]+d[i_2]$ is minimized.

Reviewer 2: ```suggestion
The problem now reduces to choosing two indices $i_1$ and $i_2$ such that $d[i_1]+d[i_2]$ is minimized.
```

However, **the two pairs cannot share any number**, so $i_1$ and $i_2$ cannot be adjacent
to each other (if $i_2=i_1+1$, the pairs would be $i_1, i_1+1$ and $i_1+1, i_1+2$, which overlap).

Without loss of generality, $i_1 < i_2$.
We loop over $2\leq i_2<n-1$ (since $i_2=1$ is impossible), so $d[i_2]$ is fixed;
$i_1$ must be in $1\leq i_1<i_2-1$, so we maintain a prefix minimum of $d[i_1]$ cumulatively

> to obtain the best $d[i_1]$ for each $i_2$.

Reviewer 2: ```suggestion
$i_1$ must be in $1\leq i_1<i_2-1$, so we maintain a prefix minimum of $d[i_1]$ to obtain the best $d[i_1]$ for each $i_2$.
```

Author: The next suggestion overode this one.

> to obtain the best $d[i_1]$ for each $i_2$.

Reviewer 2: ```suggestion
We loop over $2\leq i_2<n-1$ (since $i_2=1$ is impossible). By fixing $d[i_2], i_1$ must be in $1\leq i_1<i_2-1$, and we can maintain a prefix minimum of $d[i_1]$ to obtain the best $d[i_1]$ for each $i_2$.
```

Author: Looks good!

---
id: cfgym-105822B
source: CF
title: Kites
author: John Zhou
---

[Official Analysis](https://codeforces.com/gym/105822/attachments/download/31492/MITIT_2025_Spring_Editorials_1.pdf#page=2)

## Explanation

We must obtain $2$ pairs of the same number;
to make two indices $i_1$ and $i_2$ the same, $|a[i_1]-a[i_2]|$ operations are needed,

> so it is optimal to choose an adjacent pair of numbers to make them the same.

Reviewer 3: you say adjacent pair of numbers shouldn't it be adjacent AFTER sorting and not before?

Author: ```suggestion
We must obtain $2$ pairs of the same number;
to make two indices $i_1$ and $i_2$ the same, $|a[i_1]-a[i_2]|$ operations are needed,
so it is optimal to choose an adjacent pair of numbers in indices after sorting to make them the same.
```

Thus, we sort the input array, and compute $d[i]=a[i+1]-a[i]$ for $0<i<n-1$.
The problem now reduces to choosing two $i$'s $i_1$ and $i_2$ such that $d[i_1]+d[i_2]$ is minimized.
However, **the two pairs cannot share any number**, so $i_1$ and $i_2$ cannot be adjacent
to each other (since if $i_2=i_1+1$ or vice versa, the first pair would be $i_1, i_1+1$ and

> the second pair would be $i_1+1, i_2+1$, and $i_1+1$ is shared as an index).

Reviewer 3: no need to elaborate on this lol

Author: @Sosuke23 wanted this... I can tighten tough
```suggestion
However, **the two pairs cannot share any number**, so $i_1$ and $i_2$ cannot be adjacent
to each other (if $i_2=i_1+1$, the pairs would be $i_1, i_1+1$ and $i_1+1, i_1+2$, which overlap).
```

Reviewer 3: > @Sosuke23 wanted this... I can tighten tough

oh lol is it rlly needed? imo a silver knows THAT much

Author: yeah but i kept it to avoid risk
